File: sql-insight-engine/sql-insight-engine-innodb/src/main/java/org/gongxuanzhang/sql/insight/core/engine/storage/innodb/factory/FileHeaderFactory.java
package org.gongxuanzhang.sql.insight.core.engine.storage.innodb.factory;

import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.FileHeader;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.PageType;

import java.nio.ByteBuffer;

/**
 * @author gongxuanzhangmelt@gmail.com
 **/
public abstract class FileHeaderFactory {


    public static FileHeader readFileHeader(byte[] arr) {
        ConstantSize.FILE_HEADER.checkSize(arr);
        FileHeader fileHeader = new FileHeader();
        ByteBuffer buffer = ByteBuffer.wrap(arr);
        fileHeader.setCheckSum(buffer.getInt());
        fileHeader.setOffset(buffer.getInt());
        fileHeader.setPageType(buffer.getShort());
        fileHeader.setPre(buffer.getInt());
        fileHeader.setNext(buffer.getInt());
        fileHeader.setLsn(buffer.getLong());
        fileHeader.setFlushLsn(buffer.getLong());
        fileHeader.setSpaceId(buffer.getInt());
        return fileHeader;
    }

    /**
     * create a empty file header
     **/
    public static FileHeader createFileHeader() {
        FileHeader fileHeader = new FileHeader();
        fileHeader.setNext(0);
        fileHeader.setPre(0);
        fileHeader.setOffset(0);
        fileHeader.setPageType(PageType.FIL_PAGE_INDEX.getValue());
        return fileHeader;
    }
}

File: sql-insight-engine/sql-insight-engine-innodb/src/main/kotlin/tech/insight/engine/innodb/factory/FileHeaderFactory.kt
package tech.insight.engine.innodb.factory

import tech.insight.engine.innodb.page.ConstantSize
import tech.insight.engine.innodb.page.FileHeader
import tech.insight.engine.innodb.page.PageType
import java.nio.ByteBuffer

/**
 * @author gongxuanzhangmelt@gmail.com
 */
object FileHeaderFactory {
    fun readFileHeader(arr: ByteArray?): FileHeader {
        ConstantSize.FILE_HEADER.checkSize(arr!!)
        val fileHeader = FileHeader()
        val buffer = ByteBuffer.wrap(arr)
        fileHeader.checkSum = buffer.getInt()
        fileHeader.offset = buffer.getInt()
        fileHeader.pageType = buffer.getShort()
        fileHeader.pre = buffer.getInt()
        fileHeader.next = buffer.getInt()
        fileHeader.lsn = buffer.getLong()
        fileHeader.flushLsn = buffer.getLong()
        fileHeader.spaceId = buffer.getInt()
        return fileHeader
    }

    /**
     * create a empty file header
     */
    fun createFileHeader(): FileHeader {
        val fileHeader = FileHeader()
        fileHeader.next = 0
        fileHeader.pre = 0
        fileHeader.offset = 0
        fileHeader.pageType = PageType.FIL_PAGE_INDEX.value
        return fileHeader
    }
}
